src/v/config/constraint_types.cc

```diff
@@ -0,0 +1,176 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "config/constraint_types.h"
+
+#include "cluster/types.h"
+#include "config/constraint_store.h"
+
+namespace config {
+std::string_view to_string_view(constraint_type type) {
+    switch (type) {
+    case constraint_type::none:
+        return "none";
+    case constraint_type::restrikt:
+        return "restrict";
+    case constraint_type::clamp:
+        return "clamp";
+    }
+}
+
+template<>
+std::optional<constraint_type>
+from_string_view<constraint_type>(std::string_view sv) {
+    return string_switch<std::optional<constraint_type>>(sv)
+      .match("none", constraint_type::none)
+      .match("restrict", constraint_type::restrikt)
+      .match("clamp", constraint_type::clamp);
+}
+
+std::ostream& operator<<(std::ostream& os, const constraint_args& args) {
+    fmt::print(
+      os,
+      "{{name: {}, type: {}, min: {}, max: {}, enabled: {}}}",
+      args.name,
+      config::to_string_view(args.type),
+      args.min,
+      args.max,
+      args.enabled);
+    return os;
+}
+
+constraint_methods::constraint_methods()
+  : validator{std::move(constraint_validator_always_true)}
+  , clamper{std::move(constraint_clamper_noop)}
```

Comment:
The constructor is so more code is in the impl file instead of the header but I'm also not consistent with that so I will go back to member initialization.